sql/src/main/java/io/crate/planner/operators/LogicalPlanner.java

```diff
@@ -111,10 +112,19 @@ public LogicalPlan plan(QueriedRelation queriedRelation,
         QueriedRelation relation = optimizingRewriter.optimize(queriedRelation, plannerContext.transactionContext());
 
         LogicalPlan logicalPlan = plan(relation, fetchMode, subqueryPlanner, true)
-            .build(tableStats, new HashSet<>(relation.outputs()))
-            .tryCollapse();
+            .build(tableStats, new HashSet<>(relation.outputs()));
+
+        LogicalPlan optimizedPlan = tryOptimize(logicalPlan);
 
-        return MultiPhase.createIfNeeded(logicalPlan, relation, subqueryPlanner);
+        return MultiPhase.createIfNeeded(optimizedPlan, relation, subqueryPlanner);
+    }
+
+    private static LogicalPlan tryOptimize(LogicalPlan plan) {
+        LogicalPlan optimizedPlan = plan.tryOptimize(null);
+        if (optimizedPlan == null) {
```

Comment:
is it somewhere needed that `tryOptimize` is nullable? cant we just return the current LogicalPlan if nothing can be optimized?

Reply:
`null` is used to indicate the optimization did not succeed (i.e. an operator doesn't allow a plan to be pushed down). Since any nested plan can do what it wants, we don't know that upfront. If we return `this`, we might incorrectly optimize because the upstream operator doesn't know if the returned plan is valid or not.

The root node is not difference in this regard and that's why we have to check here although most of the time we will receive a non-null value.